pkg/ring/model.go

```diff
@@ -168,28 +202,106 @@ func (d *Desc) Ready(now time.Time, heartbeatTimeout time.Duration) error {
 	return nil
 }
 
+// RefreshTokenState updates the tokens in the ring with the state from the
+// provided tokens. RefreshTokenState will remove denormalised tokens for id
+// that are no longer in the list provided by tokens.
+func (d *Desc) RefreshTokenState(id string, tokens []StatefulToken, normalise bool) {
+	ing, ok := d.Ingesters[id]
+	if !ok {
+		return
+	}
+
+	if normalise {
+		ing.AddTokens(tokens)
+		d.Ingesters[id] = ing
+	} else {
```

Comment:
Hopefully https://github.com/cortexproject/cortex/pull/1809 gets merged soon (reviews welcome! :-)), so that this code can be removed.